### PR TITLE
DEX-279 History database

### PR DESCRIPTION
**Pull request by DenisSerebryansky**

### CODE REVIEW

File: dex-it/src/test/scala/com/wavesplatform/it/DockerContainerLauncher.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+class DockerContainerLauncher(imageName: String,
+                              containerName: String,
+                              env: String,
+                              containerPort: String,
+                              hostPort: Option[String] = None,
+                              imageTag: String = "latest") {
+
+  val dockerClient: DefaultDockerClient = DefaultDockerClient.fromEnv().build()
+
+  private val localhost: String = "0.0.0.0"
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/DockerContainerLauncher.scala]: Actually, this is not a [localhost](https://en.wikipedia.org/wiki/0.0.0.0)

[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/DockerContainerLauncher.scala]: Done

File: dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+      val createTablesDDL       = Source.fromInputStream(fileStream).getLines.toSeq.mkString
+      val createTablesStatement = sqlConnection.prepareStatement(createTablesDDL)
+
+      createTablesStatement.executeUpdate()
+      createTablesStatement.close()
+    }
+
+    Try { Class.forName(driver) } *> retry(10, 2000) { DriverManager.getConnection(url, postgresUser, postgresPassword) } >>= { sqlConnection =>
+      executeCreateTablesStatement(sqlConnection).map(_ => sqlConnection.close())
+    }
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: Fixed

File: dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        .as[PostgresConnection]("postgres")
+        .getConfig
+    )
+
+  import ctx._
+
+  def getOrdersCount: Long = ctx.run(querySchema[OrderRecord]("orders", _.id      -> "id").size)
+  def getEventsCount: Long = ctx.run(querySchema[EventRecord]("events", _.orderId -> "order_id").size)
+
+  case class EventShortenedInfo(orderId: String, eventType: Double, filled: Double, totalFilled: Double, status: Byte)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/OrderHistoryTestSuite.scala]: Fixed

File: dex/src/main/resources/order-history/order-history-ddl.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE IF NOT EXISTS orders (
+  id                VARCHAR (44) PRIMARY KEY,
+  sender            VARCHAR (44),
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/resources/order-history/order-history-ddl.sql]: Fixed

File: dex/src/main/resources/order-history/order-history-ddl.sql [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  sender_public_key VARCHAR (44),
+  amount_asset_id   VARCHAR (44),
+  price_asset_id    VARCHAR (44),
+  side              SMALLINT,
+  price             NUMERIC (27, 8),
+  amount            NUMERIC (27, 8),
+  timestamp         BIGINT,
+  expiration        BIGINT,
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/resources/order-history/order-history-ddl.sql]: Fixed

File: dex/src/main/resources/order-history/order-history-ddl.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  fee               NUMERIC (27, 8),
+  created           BIGINT
+);
+
+CREATE TABLE IF NOT EXISTS events (
+ order_id     VARCHAR(44),
+ event_type   SMALLINT,
+ timestamp    BIGINT,
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/resources/order-history/order-history-ddl.sql]: Fixed

File: dex/src/main/scala/com/wavesplatform/matcher/history/HistoryMessagesBatchSender.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      batchBuffer += msg
+  }
+
+  private def sendBatchAndAwait(): Unit = { sendBatch(); context.become(awaitingHistoryMessages) }
+
+  private def accumulateBuffer: Receive = {
+    case msg: M         => if (batchBuffer.size == batchEntries) sendBatch(); batchBuffer += msg
+    case StopAccumulate => sendBatchAndAwait()
+  }
+}
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/history/HistoryMessagesBatchSender.scala]: Fixed

File: dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    type R = OrderRecord
+    def createRecords(denormalizeAmountAndFee: (Long, AssetPair) => Double, denormalizePrice: (Long, AssetPair) => Double): Set[R] = {
+      val order = this.limitOrder.order
+      Set(
+        OrderRecord(
+          id = order.id().toString,
+          sender = order.sender.toString,
+          senderPublicKey = order.senderPublicKey.toString,
+          amountAssetId = order.assetPair.amountAsset.fold(AssetPair.WavesName)(_.id.toString),
+          priceAssetId = order.assetPair.priceAsset.fold(AssetPair.WavesName)(_.id.toString),
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala]: Done

File: dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+          Set(submitted -> e.submittedRemainingAmount, counter -> e.counterRemainingAmount) map {
+            case (limitOrder, remainingAmount) =>
+              EventRecord(
+                orderId = limitOrder.order.id().toString,
+                eventType = eventTrade,
+                timestamp = timestamp,
+                price = denormalizePrice(limitOrder.order.price, assetPair),
+                filled = denormalizeAmountAndFee(e.executedAmount, assetPair),
+                totalFilled = denormalizeAmountAndFee(limitOrder.order.amount - remainingAmount, assetPair),
+                status = if (remainingAmount != 0) statusPartiallyFilled else statusFilled
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala]: Done

File: dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala [hunk 10]
```diff
@@ -22,6 +22,32 @@ object MatcherModel {
   def fromNormalized(value: Long, amountAssetDecimals: Int, priceAssetDecimals: Int): Double = {
     (BigDecimal.valueOf(value) / BigDecimal(10).pow(8 + priceAssetDecimals - amountAssetDecimals).toLongExact).toDouble
   }
+
+  def denormalizeAmountAndFee(value: Long, blockchain: Blockchain, pair: AssetPair): Double = {
+    fromNormalized(value, -getAssetDecimals(blockchain, pair.amountAsset), -8)
+  }
+
+  def denormalizeAmountAndFee(value: Long, amountAssetDecimals: Byte): Double = {
+    fromNormalized(value, -amountAssetDecimals, -8)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala]: Refactored

File: dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala [hunk 11]
```diff
@@ -22,6 +22,32 @@ object MatcherModel {
   def fromNormalized(value: Long, amountAssetDecimals: Int, priceAssetDecimals: Int): Double = {
     (BigDecimal.valueOf(value) / BigDecimal(10).pow(8 + priceAssetDecimals - amountAssetDecimals).toLongExact).toDouble
   }
+
+  def denormalizeAmountAndFee(value: Long, blockchain: Blockchain, pair: AssetPair): Double = {
+    fromNormalized(value, -getAssetDecimals(blockchain, pair.amountAsset), -8)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala]: Refactored

File: dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala [hunk 12]
```diff
@@ -22,6 +22,32 @@ object MatcherModel {
   def fromNormalized(value: Long, amountAssetDecimals: Int, priceAssetDecimals: Int): Double = {
     (BigDecimal.valueOf(value) / BigDecimal(10).pow(8 + priceAssetDecimals - amountAssetDecimals).toLongExact).toDouble
   }
+
+  def denormalizeAmountAndFee(value: Long, blockchain: Blockchain, pair: AssetPair): Double = {
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/model/MatcherModel.scala]: Refactored

File: dex/src/test/scala/com/wavesplatform/matcher/history/HistoryRouterSpecification.scala [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  "HistoryRouter" should {
+    "correctly convert events to records" in {
+
+      import HistoryRouter._
+
+      // place big buy order
+      getOrderInfo(orderAdded(buyWavesOrder)) shouldBe
+        OrderShortenedInfo(buyWavesOrder.orderId, buyWavesOrder.sender, side = 0, price = 1, amount = 300)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/test/scala/com/wavesplatform/matcher/history/HistoryRouterSpecification.scala]: Done

File: project/Dependencies.scala [hunk 14]
```diff
@@ -147,8 +152,8 @@ object Dependencies {
   ) ++ Seq(
     akkaModule("testkit"),
     akkaModule("persistence-tck"),
-    "com.github.dnvriend" %% "akka-persistence-inmemory" % "2.5.15.1"
-  ).map(_ % Test) ++ test
+    "com.github.dnvriend" %% "akka-persistence-inmemory" % "2.5.15.1",
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply project/Dependencies.scala]: Done

File: dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    def createRecords(denormalizeAmountAndFee: (Long, AssetPair) => Double, denormalizePrice: (Long, AssetPair) => Double): Set[R]
+  }
+
+  final case class SaveOrder(limitOrder: LimitOrder, timestamp: Long) extends HistoryMsg {
+    type R = OrderRecord
+    def createRecords(denormalizeAmountAndFee: (Long, AssetPair) => Double, denormalizePrice: (Long, AssetPair) => Double): Set[R] = {
+      val order = this.limitOrder.order
+      Set(
+        OrderRecord(
+          id = order.id().toString,
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/matcher/history/HistoryRouter.scala]: Done

**Pull request merged**